### ISSUE TITLE
[Discussion] Loop Tiling Could possibly make it faster for clustering

### ISSUE THREAD

**Issue by andyafter** [krukah/robopoker#36]
Throughout the previous couple days I have tried various instances on AWS, from C6, C7 series to R6, R7, All the instances that I have tried have more than 96 cores and 256GB RAM.

I found that the initialization of clustering constantly remains around 5~11 hours for the kmeans initialization of flop. 

However, HPC6a.48Xlarge instances demonstrated a significant improvement in this process, with 12 minutes initialization!!!

<img width="1432" alt="Image" src="https://github.com/user-attachments/assets/bd314850-252c-4fd0-ad3d-6337d0e8cae0" />

The HPC instance that I used contains around 384GB of RAM(which was barely used), and 96cores, and most importantly, 512MB of L3 cache and 48GB of L2 cache. C6a.32xlarge has 128MB of L3 cache, that seems to be the key that caused this training time improvement. 

Which, my conclusion is that, the clustering is handling significant amount of repeatitive memory fetching and writing, with which we could implement a loop tiling technique to significantly boost the clustering efficiency. 

What do you guys think!

**Comment by krukah** [comment]
this sounds quite promising! it is certainly true that we do a ton of repetitive memory access in the training steps, and i've not thought about how to optimize these operations. what do you imagine a loop tiling approach could look like here?

**Comment by andyafter** [comment]
Still trying to get a good understand of the clustering code. Let me answer this a bit later.

**Comment by andyafter** [comment]
> this sounds quite promising! it is certainly true that we do a ton of repetitive memory access in the training steps, and i've not thought about how to optimize these operations. what do you imagine a loop tiling approach could look like here? 

Perhaps, if I'm understanding everything correctly, if you could help explain a bit on the logic of the clustering and postgresql formatting here, I could help write this clustering with tensor flow, maybe getting a faster training speed.

**Comment by krukah** [comment]
1. postgres binary format is sorta arbitrary, but the format is described here. we could use any serialization method, but i chose postgres just because it makes uploading into an indexed database much easier.
2. the main data structures involved in clustering are
- `Histogram` which is implemented as a `BTreeMap<Abstraction, Probability(f32)>` but more generally, the important trait that it implements is [`Density`](https://github.com/krukah/robopoker/blob/0abc70667bed1225c22b26cf3c04e1d77fe7acce/src/transport/density.rs). however, if the BTreeMap becomes a problem (since an equivalent doesn't exist in TensorFlow afaik?), we can implement this interface with a Vec<Abstraction> or even [Abstraction; N]
- `Metric` which is implemented as a `BTreeMap<Pair, f32>`. this object should be pretty (relatively) small (n_abstractions choose 2 ~ 10k) and should ideally always be in some memory cache, since every EMD calculation will reference it. not sure how to represent this in TensorFlow though.

i'm not sure if this helps, but i can answer specific questions around any part of the clustering code.

generally, i think some speedup from "fixed-length array in tensorFlow or other lin alg library" is possible, and it would require (approximately) the following implemenation refactors:
- `Histogram(BTreeMap<Abstraction, f32>) -> Histogram([Abstraction; N_FLOP_CLUSTERS])`
- possibly a refactor of the current EMD implementation which uses Sinkhorn algorithm

**Comment by krukah** [comment]
https://docs.google.com/document/d/1VGQmarcjzcR378lBzvyb9A1tcW-Hd1_SuIGxaNpn3fQ/edit?tab=t.0
came back around to answerting these questions, they're good questions so i think i'll post them here for posterity
